### PR TITLE
Introduces the `misk-crypto-testing` module

### PR DESCRIPTION
**Pull request by jason-stiefel**
Externalization of the crypto test fixture allows for more flexible downstream usage without importing things like `FakeMetrics` (relevant in subsequent PR) 

- Deprecates existing `CryptoTestModule` and copies implementation into `misk-crypto-testing`
- Adds `.internal` packages for `Provider<KeyType>` definitions and opens the classes up for wiring by `CryptoTestModule`

### CODE REVIEW

File: misk-crypto/src/main/kotlin/misk/crypto/KeyReader.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  private fun readEncryptedKey(key: Key): KeysetHandle {
+    val masterKey = kmsClient.getAead(key.kms_uri)
+    return try {
+      val kek = KmsEnvelopeAead(KEK_TEMPLATE, masterKey)
+      val reader = JsonKeysetReader.withString(key.encrypted_key!!.value)
+      KeysetHandle.read(reader, kek)
+    } catch (ex: GeneralSecurityException) {
+      logger.warn { "using obsolete key format, rotate your keys when possible" }
```
[the remark this answers is not in the document]
Comment (yoavamit) [reply misk-crypto/src/main/kotlin/misk/crypto/KeyReader.kt:40]: Yeah, might be a good idea to start shedding some of this old code and clean in it up.

**Review by yoavamit — COMMENTED** [review]
This looks very promising.

I'd take another pass at the code and make sure it's aligned with the rest of the project.
Additionally, could you please add in the PR description what is the motivation for this change?

File: misk-crypto-testing/src/main/kotlin/misk/crypto/testing/CryptoTestModule.kt [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    val externalDataKeys = config.external_data_keys ?: emptyMap()
+    bind(object : TypeLiteral<Map<KeyAlias, KeyType>>() {})
+      .annotatedWith(ExternalDataKeys::class.java)
+      .toInstance(externalDataKeys)
+    keyManagerBinder.addBinding().toInstance(FakeKeyResolver(externalDataKeys))
+
+    if (externalDataKeys.isNotEmpty()) {
+      externalDataKeys.entries.forEach { entry ->
+        val fakeFake = Key(entry.key, entry.value, MiskConfig.RealSecret(""))
```
[the remark this answers is not in the document]
Comment (jason-stiefel) [reply misk-crypto-testing/src/main/kotlin/misk/crypto/testing/CryptoTestModule.kt:84]: Just want to point out that line points to you in `blame` :-D

File: misk-crypto-testing/src/main/kotlin/misk/crypto/testing/FakeKeyResolver.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  constructor(rawKeys: List<Key>) {
+    rawKeys.forEach { key ->
+      val plaintextSecret = keyTypeToSecret(key.key_type)
+      val encryptedSecret = if (
+        key.key_type == KeyType.HYBRID_ENCRYPT ||
+        key.key_type == KeyType.PGP_DECRYPT ||
+        key.key_type == KeyType.PGP_ENCRYPT
+      ) {
+        plaintextSecret
+      } else if (key.key_name == "obsolete") {
```
[the remark this answers is not in the document]
Comment (jason-stiefel) [reply misk-crypto-testing/src/main/kotlin/misk/crypto/testing/FakeKeyResolver.kt:37]: Serves a wildly obfuscated purpose to this test: https://github.com/cashapp/misk/blob/e7db88dbbdfa003632bc2e2320e638766ad02709/misk-crypto-testing/src/test/kotlin/misk/crypto/CryptoModuleTest.kt#L184

File: misk-crypto-testing/src/main/kotlin/misk/crypto/testing/FakeKeyResolver.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        key.encrypted_key
+      } else {
+        TestKeysets.encryptSecret(plaintextSecret)
+      }
+      if (key.key_type == KeyType.HYBRID_ENCRYPT) {
+        returnedKeysets[key.key_name] = key.copy(encrypted_key = encryptedSecret, kms_uri = null)
+      } else {
+        returnedKeysets[key.key_name] = key.copy(encrypted_key = encryptedSecret, kms_uri = "stub")
```
[the remark this answers is not in the document]
Comment (jason-stiefel) [reply misk-crypto-testing/src/main/kotlin/misk/crypto/testing/FakeKeyResolver.kt:45]: This depends on the use of `FakeKmsClient` thats bound in `CryptoTestModule`.

File: misk-crypto-testing/src/main/kotlin/misk/crypto/testing/FakeKeyResolver.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        throw UnsupportedOperationException("PGP keys are not supported yet as external keys")
+      }
+      val secret = keyTypeToSecret(type)
+      returnedKeysets[alias] = Key(alias, type, secret)
+    }
+  }
+
+  // Mock local keys
+  constructor(rawKeys: List<Key>) {
+    rawKeys.forEach { key ->
```
[the remark this answers is not in the document]
Comment (jason-stiefel) [reply misk-crypto-testing/src/main/kotlin/misk/crypto/testing/FakeKeyResolver.kt:29]: Tinkered with this for a bit and got mucked up between the 2 constructors - if you have time lets pair it on Monday and clean it up?

**Review by yoavamit — APPROVED** [review]

**Pull request merged**